api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -232,7 +263,59 @@ message DeltaDiscoveryResponse {
   config.core.v3.ControlPlane control_plane = 7;
 }
 
-// [#next-free-field: 8]
+// A set of dynamic parameter constraints associated with an individual
+// xDS resource. These constraints determine whether the resource matches
+// a subscription based on the set of dynamic parameters in the subscription.
+// This allows xDS servers and caching proxies to determine which variant of a
+// resource is appropriate for a given client.
+message DynamicParameterConstraints {
+  message Constraint {
+    // A list of one or more integer ranges.
+    // A value is considered to match if it falls in any of the ranges.
+    message IntegerRangeList {
+      // At least one of *min_value* or *max_value* must be set.
+      message Range {
+        // If specified, value may not be less than this.
+        uint64 min_value = 1;
+
+        // If specified, value may not be greater than this.
+        uint64 max_value = 2;
+      }
+
+      repeated Range range = 1;
+    }
+
+    message Exists {
+    }
+
+    oneof constraint {
+      // The key must have this specific value.
+      string value = 1;
+
+      // The key's value must be integers and within one of the ranges in this list.
+      IntegerRangeList integer_range_list = 2;
+
+      // The key exists.
+      Exists exists = 3;
+    }
+
+    // If true, the constraint is inverted (i.e., logical-not of the constraint).
+    bool invert = 4;
+  }
+
+  // The name of the dynamic parameter to match against.
+  string key = 1;
+
+  // A list of one or more constraints on the value of the key.
+  // All constraints must be met.
+  repeated Constraint constraints = 2;
```

Comment:
Yet more matching. :) Should we be using the generic matching framework for this? (Since I almost guarantee someone is going to want OR semantics, etc. cc @snowp

Reply:
I think this would be an interesting use case of it, as iiuc this would get sent to the server and have to get resolved there? So while the API could be reused we'd have to implement the matching engine in all the languages trying to make use of this (which I personally wouldn't mind, would make testing matching config in Go etc. a lot easier :) )

Reply:
As it happens, we may wind up adding some simple OR semantics here.  I discussed this with @fuqianggao the other day, and he's going to try to gather some data from our production configs to tell us whether that's something we really need.

I did actually consider using the generic matching API, but I wound up deciding against it for the following reasons:
- The constraints specified here need to be part of the cache key for each resource, and one of the drawbacks of this design is that it's more difficult to optimize the cache implementation when the cache key involves a set of constraints than it is when the cache key just contains a set of fixed key/value pairs.  We are introducing some complexity here for caches, and I'd prefer to keep that to a minimum.
- The generic matching API is intended to be very flexible, with a lot of extension points that make it easy to inject new functionality.  However, the matching semantics here need to be understood by every xDS client and server, which means that if/when we need to add new types of constraints here, we'll probably need to add a new client capability, so that the server knows that it's safe to send the new types of constraints.  Given that that's going to be a fairly heavyweight operation anyway, the extensibility of the generic matching API doesn't seem to provide much benefit.
- This type of matching will need to be supported by every xDS client and server at the transport protocol level, and the machinery for the generic matching API seems like a fairly large dependency to drag into the transport protocol.

All of that having been said, there is definitely a trade-off to be made here.  I'm certainly open to arguments that we should be trying to be more flexible here.